feat(language_server): search for nested configurations by initialization

Sysix opened the pull request:
Now the client does not need to send a second request after initialization.
The initialization part is done on the server side. The client only needs to send future changes because file-watchers on servers are not recommended: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles

Sysix commented:
* **#10120** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10120?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10120?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10120).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented:
fyi @nrayburn-tech @camchenry  
![filewatcher-duration](https://github.com/user-attachments/assets/0b4645ca-1be9-49ab-864d-3c0a35709e2d)

Sysix commented on editors/vscode/client/extension.ts:
```diff
@@ -297,52 +287,28 @@ export async function activate(context: ExtensionContext) {
+  client = undefined;
 }
 
+// FileSystemWatcher are not ready on the start and can take some seconds on bigger repositories
```
That is the relevant part of the gif

Sysix replied:
Looks like this is a "me" problem: https://github.com/microsoft/vscode/issues/245211

camchenry commented on crates/oxc_language_server/src/linter/config_walker.rs:
```diff
@@ -0,0 +1,87 @@
+        }
+        let Some(file_name) = entry.path().file_name() else { return false };
+
+        file_name == ".oxlintrc.json"
```
do we have this in a const somewhere?

camchenry reviewed (approved):
looks good, I do wonder if we should use the config walker in oxlint too. not sure how it'd affect perf, but it would make the implementation across both consistent which would be nice.

Sysix commented:
> looks good, I do wonder if we should use the config walker in oxlint too. not sure how it'd affect perf, but it would make the implementation across both consistent which would be nice.

Yes, I was close to refactor it already but using 2x a Walker  sounds not good for perf. Maybe we can combine them into one?

Sysix commented:
### Merge activity

* **Apr 4, 7:29 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Sysix committed:
feat(language_server): search for nested configurations by initialization (#10120)

Now the client does not need to send a second request after initialization.
The initialization part is done on the server side. The client only needs to send future changes because file-watchers on servers are not recommended: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles
